testng-core/src/main/java/org/testng/internal/invokers/TestMethodWorker.java

```diff
@@ -119,8 +121,9 @@ && doesTaskHavePreRequisites()
 
     for (IMethodInstance testMethodInstance : m_methodInstances) {
       ITestNGMethod testMethod = testMethodInstance.getMethod();
+      String key = Integer.toString(testMethod.getInstance().toString().hashCode());
```

Comment:
`hashCode` is not unique. Please do not use it for identity

Reply:
I am getting the `hashCode()` of the string representation of an object (via `toString()`). I couldn't find any other ways of getting a shorter key. Please let me know if there are any other alternatives.

Reply:
I did a quick dirty test to check if the string's `hashCode()` is impacted by `-XX:+UnlockExperimentalVMOptions -XX:hashCode=2` It looks like the impact is perhaps only on user defined objects? So me using the `hashCode()` on  `toString()` should be fine if at all the user doesn't override the `toString()` of objects to return the same value because if they do that, then I am back to square one and not sure how to deal with this situation. I currently am not able to use the `testClass` object as a key because then it prevents it from being gc'ed and the build fails since we have a test case that vets out this gc part.

Reply:
I think we need to overhaul TestNG in such a way that we have a container object which has an object id and the actual object by itself which we track as soon as we create an instance. That way we will be able to use the object id (perhaps from a UUID for e.g.,) But that will be a bit more elaborate change. I will want to handle that in a separate PR. Thoughts @vlsi @juherr ?

Reply:
`toString` can be overridden the same as `hashCode`. A maybe better solution will be [identityHashCode](https://docs.oracle.com/javase/6/docs/api/java/lang/System.html#identityHashCode%28java.lang.Object%29) but not always unique.

As I understand it, `KeyAwareAutoCloseableLock.AutoReleasable` is currently broken by design and cannot be used without a deeper change.
I don't know what is the best solution for that change. Maybe using a holder object with an UUID and its object could be a solution.

Reply:
@juherr - `System.identityHashCode()` also suffers from the same problem of giving the same hashCode if the user overridden hashCode is either poor by choice wherein it generates clashes or when the user ends up passing `-XX:+UnlockExperimentalVMOptions -XX:hashCode=2` which causes the hashCode to be the same across the board. 

Here's a sample that uses the above JVM flags along with its output

```java
public class StringGames {

    public static void main(String[] args) {
        String text = "Hello";
        Integer integer = 1;
        System.err.println("text.hashCode() =====>" + text.hashCode());
        System.err.println("System.identityHashCode(text) =====>" + System.identityHashCode(text));
        System.err.println("System.identityHashCode(new Book()) =====>" + System.identityHashCode(new Book()));
        System.err.println("new Book().toString().hashCode() =====>" + new Book().toString().hashCode());
        System.err.println("integer.hashCode() =====>" + integer.hashCode());
        System.err.println("System.identityHashCode(integer) =====>" + System.identityHashCode(integer));
    }

    public static class Book {}
}
```

```bash
text.hashCode() =====>69609650
System.identityHashCode(text) =====>1
System.identityHashCode(new Book()) =====>1
new Book().toString().hashCode() =====>-246658450
integer.hashCode() =====>1
System.identityHashCode(integer) =====>696476745
```


I think we should be able to get this PR merged with the current changes. I will work on raising a new PR wherein we enhance TestNG to start using a wrapper object which contains a `UUID` for every new instance instantiated by TestNG. Once that change is done, I can come back and refactor the code to start using the `UUID` associated with the instance as the key and thus the implementation would be more predictable.

Reply:
@juherr - Here's the issue that I created for this https://github.com/testng-team/testng/issues/3079

Reply:
Ok. Then rollback `KeyAwareAutoCloseableLock.AutoReleasable` and add it into the #3079 fix

Reply:
So that would mean that only this section would continue to use the synchronized keyword. 

Is that what you are suggesting?

Reply:
Yes. There is no reason to introduce a bug.

Reply:
@juherr Fixed. Can we merge this ? (I would like to squash the commits and merge them)